pkg/controller/packetsampling/controller.go

```diff
@@ -0,0 +1,359 @@
+	return err
+}
+
+// Allocates a tag. If the PacketSampling request has been allocated with a tag
+// already, 0 is returned. If number of existing PacketSampling requests reaches
+// the upper limit, an error is returned.
+func (c *Controller) allocateTag(name string) (uint8, error) {
+	c.runningPacketSamplingsMutex.Lock()
+	defer c.runningPacketSamplingsMutex.Unlock()
+
+	for _, n := range c.runningPacketSamplings {
+		if n == name {
+			// The packetsampling request has been processed already.
+			return 0, nil
+		}
+	}
+	for i := minTagNum; i <= maxTagNum; i += tagStep {
+		if _, ok := c.runningPacketSamplings[i]; !ok {
+			c.runningPacketSamplings[i] = name
+			return i, nil
+		}
+	}
+	return 0, fmt.Errorf("number of on-going packetsampling operations already reached the upper limit: %d", maxTagNum)
+}
+
+func (c *Controller) processPacketSamplingItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer c.queue.Done(obj)
+
+	key, ok := obj.(string)
+	if !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	}
+
+	err := c.syncPacketSampling(key)
+
+	if err != nil {
+		klog.Errorf("error sync packetSampling %s, existing,error: %v", key, err)
+		c.queue.AddRateLimited(key)
+	} else {
+		c.queue.Forget(key)
+	}
+
+	return true
+}
+
+func (c *Controller) deallocateTagForPS(ps *crdv1alpha1.PacketSampling) {
+	if ps.Status.DataplaneTag != 0 {
+		c.deallocateTag(ps.Name, uint8(ps.Status.DataplaneTag))
+	}
+}
+
+func (c *Controller) deallocateTag(name string, tag uint8) {
+	c.runningPacketSamplingsMutex.Lock()
+	defer c.runningPacketSamplingsMutex.Unlock()
+
+	if existing, ok := c.runningPacketSamplings[tag]; ok {
+		if name == existing {
+			delete(c.runningPacketSamplings, tag)
+		}
+	}
+}
+
+func (c *Controller) occupyTag(ps *crdv1alpha1.PacketSampling) error {
+	tag := uint8(ps.Status.DataplaneTag)
+
+	if tag < minTagNum || tag > maxTagNum {
+		return errors.New("this packetsamping crd's data plane tag is out of range")
+	}
+	c.runningPacketSamplingsMutex.Lock()
+	defer c.runningPacketSamplingsMutex.Unlock()
+
+	if exist, ok := c.runningPacketSamplings[tag]; ok {
+		if ps.Name == exist {
+			return nil
+		}
+		return errors.New("this packetsamping crd's data plane tag is already occupied")
+	}
+	c.runningPacketSamplings[tag] = ps.Name
+	return nil
+
+}
+
+func (c *Controller) syncPacketSampling(name string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished sync for packetsampling %s.(%v)", name, time.Since(startTime))
+	}()
+
+	ps, err := c.packetSamplingLister.Get(name)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	switch ps.Status.Phase {
+	case "":
+		err = c.startPacketSampling(ps)
+	case crdv1alpha1.PacketSamplingRunning:
+		err = c.checkPacketSamplingStatus(ps)
+	case crdv1alpha1.PacketSamplingFailed:
```

Comment:
I suppose you should also deallocate the tag if the phase is succeed?

Reply:
it's already included in the `checkPacketSamplingStatus` method.